### PR TITLE
Timeout and disconnect improvements

### PR DESCRIPTION
**Pull request by kberg-amzn**

### CODE REVIEW

File: Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpConnection.h [hunk 1]
```diff
@@ -136,19 +136,19 @@ namespace AzNetworking
...
+        ConnectionRole   m_connectionRole = ConnectionRole::Connector;
+        DtlsEndpoint     m_dtlsEndpoint;
 
         AZ::TimeMs m_lastSentPacketMs;
         uint32_t   m_unackedPacketCount = 0;
         uint32_t   m_connectionMtu = MaxUdpTransmissionUnit;
 
         TimeoutId m_timeoutId;
-        uint32_t  m_timeoutCounter = 0;
+        int32_t   m_timeoutCounter = 0;
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpConnection.h:151]: Why the change from uint to int?

File: Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@ namespace AzNetworking
 
     AZ_CVAR(bool, net_UdpTimeoutConnections, true, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "Boolean value on whether we should timeout Udp connections");
     AZ_CVAR(AZ::TimeMs, net_UdpPacketTimeSliceMs, AZ::TimeMs{ 8 }, nullptr, AZ::ConsoleFunctorFlags::DontReplicate, "The number of milliseconds to allow for packet processing");
-    AZ_CVAR(AZ::TimeMs, net_UdpHeartbeatTimeMs, AZ::TimeMs{ 2 * 1000 }, nullptr, AZ::ConsoleFunctorFlags::Null, "Udp connection heartbeat frequency");
+    AZ_CVAR(int32_t, net_UdpUnackedHeartbeats, 5, nullptr, AZ::ConsoleFunctorFlags::Null, "The number of heartbeats to attempt to send to keep a connection alive before giving up");
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpNetworkInterface.cpp:34]: uint to prevent negative values would be good here

File: Gems/Multiplayer/Code/Source/NetworkEntity/NetworkEntityAuthorityTracker.h [hunk 3]
```diff
@@ -29,37 +29,13 @@ namespace Multiplayer
...
-            NetworkEntityTimeoutFunctor(NetworkEntityAuthorityTracker& networkEntityAuthorityTracker, INetworkEntityManager& m_networkEntityManager);
-            AzNetworking::TimeoutResult HandleTimeout(AzNetworking::TimeoutQueue::TimeoutItem& item) override;
-        private:
-            AZ_DISABLE_COPY_MOVE(NetworkEntityTimeoutFunctor);
-            NetworkEntityAuthorityTracker& m_networkEntityAuthorityTracker;
-            INetworkEntityManager& m_networkEntityManager;
-        };
-
-        using TimeoutDataMap = AZStd::unordered_map<NetEntityId, TimeoutData>;
+        using TimeoutDataMap = AZStd::unordered_set<NetEntityId>;
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/NetworkEntity/NetworkEntityAuthorityTracker.h:34]: TimeoutDataMap and m_timeoutDataMap are pretty confusing name wise now. Maybe TimeoutNetEntities and m_timeoutNetIds?

File: Gems/Multiplayer/Code/Include/Multiplayer/EntityDomains/IEntityDomain.h [hunk 4]
```diff
@@ -34,12 +32,10 @@ namespace Multiplayer
...
 
-        //! Enable Entity Domain Exit Tracking for entities on the host.
-        //! @param ownedEntitySet the set of entities to activate tracking for
-        virtual void ActivateTracking(const INetworkEntityManager::OwnedEntitySet& ownedEntitySet) = 0;
-
-        //! Return the set of netbound entities not included in this domain.
-        virtual const EntitiesNotInDomain& RetrieveEntitiesNotInDomain() const = 0;
+        //! This method will be invoked whenever we unexpectedly lose the authoritative entity replicator for an entity.
+        //! This gives our entity domain a chance to determine whether or not it should assume authority in this instance.
+        //! @param entityHandle the network entity handle of the entity that has lost it's authoritative replicator
```
Comment (AMZN-Gene) [inline Gems/Multiplayer/Code/Include/Multiplayer/EntityDomains/IEntityDomain.h:37]: Remove apostrophe, “lost its…”

**Review by AMZN-Gene — APPROVED** [review]

**Review by SergeyAMZN — APPROVED** [review]

File: Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpFragmentQueue.cpp [hunk 5]
```diff
@@ -20,7 +20,13 @@ namespace AzNetworking
 
     void UdpFragmentQueue::Update()
     {
-        m_timeoutQueue.UpdateTimeouts(*this);
+        m_timeoutQueue.UpdateTimeouts([this](TimeoutQueue::TimeoutItem& item)
+        {
+            const SequenceId fragmentSequence = static_cast<SequenceId>(item.m_userData & 0xFF);
```
Comment (AMZN-puvvadar) [inline Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpFragmentQueue.cpp:25]: NP: We should const the mask so people know what each chunk of bits represents

**Review by AMZN-puvvadar — APPROVED** [review]
